ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -0,0 +1,647 @@
+import           Prelude hiding (words)
+import qualified System.Random.Stateful as R
+import qualified Test.Ouroboros.Consensus.ChainGenerator.BitVector as BV
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Counting as C
+import           Test.Ouroboros.Consensus.ChainGenerator.Params (Asc,
+                     Delta (Delta), Kcp (Kcp), Len (Len), Scg (Scg), ascVal)
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Slot as S
+import           Test.Ouroboros.Consensus.ChainGenerator.Slot
+                     (E (ActiveSlotE, SlotE), S)
+import qualified Test.Ouroboros.Consensus.ChainGenerator.Some as Some
+
+-----
+
+data HonestRecipe = HonestRecipe !Kcp !Scg !Delta !Len
+  deriving (Eq, Read, Show)
+
+data CheckedHonestRecipe base hon = UnsafeCheckedHonestRecipe {
+    chrEhcgDensity :: !(BV.SomeDensityWindow S.NotInverted)
+  ,
+    chrWin         :: !(C.Contains SlotE base hon)
+  }
+  deriving (Eq, Read, Show)
+
+data SomeCheckedHonestRecipe =
+    forall base hon.
+    SomeCheckedHonestRecipe
+        !(Proxy base)
+        !(Proxy hon)
+        !(CheckedHonestRecipe base hon)
+
+instance Show SomeCheckedHonestRecipe where
+    showsPrec p (SomeCheckedHonestRecipe base hon recipe) =
+        Some.runShowsPrec p
+      $ Some.showCtor SomeCheckedHonestRecipe "SomeCheckedHonestRecipe"
+            `Some.showArg` base
+            `Some.showArg` hon
+            `Some.showArg` recipe
+
+instance Read SomeCheckedHonestRecipe where
+    readPrec =
+        Some.runReadPrec
+      $ Some.readCtor SomeCheckedHonestRecipe "SomeCheckedHonestRecipe"
+            <*> Some.readArg
+            <*> Some.readArg
+            <*> Some.readArg
+
+data NoSuchHonestChainSchedule =
+    -- | 'Scg' must be greater than 'Delta'
+    BadDeltaScg
+  |
+    -- | must have @0 <= 'Kcp' < 'Scg' - 'Delta'@
+    BadKcp
+  |
+    -- | 'Len' must be positive
+    BadLen
+  deriving (Eq, Read, Show)
+
+checkHonestRecipe :: HonestRecipe -> Exn.Except NoSuchHonestChainSchedule SomeCheckedHonestRecipe
+checkHonestRecipe recipe = do
+    when (s <= d) $ Exn.throwError BadDeltaScg
+
+    when (l <= 0) $ Exn.throwError BadLen
+
+    when (k < 0 || s - d <= k) $ Exn.throwError BadKcp
+
+    C.withTopWindow (C.Lbl @HonestLbl) l $ \base topWindow -> do
+        C.SomeWindow Proxy slots <- pure topWindow
+
+        pure $ SomeCheckedHonestRecipe base Proxy UnsafeCheckedHonestRecipe {
+            chrEhcgDensity = BV.SomeDensityWindow (C.Count (k + 1)) (C.Count (s - d))
+          ,
+            chrWin         = slots
+          }
+  where
+    HonestRecipe (Kcp k) (Scg s) (Delta d) (Len l) = recipe
+
+-----
+
+-- | The leader schedule of an honest /chain/
+--
+-- The represented chain grows by one block per active slot. A different data
+-- type would represent a leader schedule in which leaders to not extend a block
+-- from the previous active slot.
+--
+-- INVARIANT: at least one active slot
+data ChainSchedule base inner =
```

Comment:
Maybe `ChainSchedule` needs to be renamed to `HonestChainSchema` to follow the test plan terminology.

Reply:
I agree 100%

Reply:
Renamed ChainSchedule to ChainSchema in bb2d522bf. I could not prefix the name with "Honest" because the PR seems to use the same type for the adversarial schema.